doc/fwutil/fwutil.md

```diff
@@ -0,0 +1,441 @@
+|:---:|:----------:|:--------------:|:----------------------------------|
+| 0.1 | 21/08/2019 | Nazarii Hnydyn | Initial version                   |
+| 0.2 | 10/09/2019 | Nazarii Hnydyn | Review feedback and other changes |
+| 0.3 | 17/09/2019 | Nazarii Hnydyn | Align flows with the platform API |
+
+## Abbreviations
+
+| Term   | Meaning                                             |
+|:-------|:----------------------------------------------------|
+| FW     | Firmware                                            |
+| SONiC  | Software for Open Networking in the Cloud           |
+| PSU    | Power Supply Unit                                   |
+| QSFP   | Quad Small Form-factor Pluggable                    |
+| EEPROM | Electrically Erasable Programmable Read-Only Memory |
+| I2C    | Inter-Integrated Circuit                            |
+| SPI    | Serial Peripheral Interface                         |
+| JTAG   | Joint Test Action Group                             |
+| BIOS   | Basic Input/Output System                           |
+| CPLD   | Complex Programmable Logic Device                   |
+| FPGA   | Field-Programmable Gate Array                       |
+| URL    | Uniform Resource Locator                            |
+| API    | Application Programming Interface                   |
+| N/A    | Not Applicable/Not Available                        |
+
+## List of figures
+
+[Figure 1: FW utility High Level Design](#figure-1-fw-utility-high-level-design)  
+[Figure 2: Show components status flow](#figure-2-show-components-status-flow)  
+[Figure 3: FW install (non modular) flow](#figure-3-fw-install-non-modular-flow)  
+[Figure 4: FW install (modular) flow](#figure-4-fw-install-modular-flow)  
+
+## List of tables
+
+[Table 1: Event logging](#table-1-event-logging)
+
+# 1 Introduction
+
+## 1.1 Feature overview
+
+A modern network switch is a sophisticated equipment which consists of many auxiliary components  
+which are responsible for managing different subsystems (e.g., PSU/FAN/QSFP/EEPROM/THERMAL)  
+and providing necessary interfaces (e.g., I2C/SPI/JTAG).
+
+Basically these components are complex programmable logic devices with it's own HW architecture  
+and software. The most important are BIOS/CPLD/FPGA etc.
+
+It is very important to always have the latest recommended software version to improve device stability,  
+security and performance. Also, software updates can add new features and remove outdated ones.
+
+In order to make software update as simple as possible and to provide a nice user frindly  
+interface for various maintenance operations (e.g., install a new FW or query current version)  
+we might need a dedicated FW utility.
+
+## 1.2 Requirements
+
+### 1.2.1 Functionality
+
+**This feature will support the following functionality:**
+1. Manual FW installation for particular platform component
+2. Automatic FW installation for all available platform components
```

Comment:
Can ONIE itself be a valid platform component ? Say, we want to upgrade ONIE using the fwutil interface : can we provide a onie-updater image to trigger the onie-self-update ?

Reply:
> 
> 
> Can ONIE itself be a valid platform component ? Say, we want to upgrade ONIE using the fwutil interface : can we provide a onie-updater image to trigger the onie-self-update ?

@padmanarayana from architecture point of view ONiE is not a platform component.
But you can consider it as platform component and use platform API for scheduling future ONiE updates.

But in this case the bootloader update might be needed since we have some additional flows for SONiC-To-SONiC update and booting to ONiE for maintenance operations can break it.

Reply:
Agree that ONIE is not a platform component per-se. We do understand the caveats - but just wanted to make sure that the fwutil does not preclude this case. Thanks for the clarifications!

Reply:
@padmanarayana welcome